### PR TITLE
Column generation for persistent solvers

### PR DESCRIPTION
**Pull request by bknueven**
## Summary/Motivation:
Column generation is a classical decomposition technique, for which Pyomo currently has no built-in support. Further, most MIP solvers support adding a column as part of their add variable method. This PR would augment the existing persistent solver interfaces to better enable writing column generation algorithms in Pyomo.

## Changes proposed in this PR:
- `add_column` method for `PersistentSolver`
- `_add_column` methods for all existing derived classes
- Slight refactoring of existing direct/persistent solvers to avoid repeating code three times (`lb_ub_from_var` methods).
- Tests of the new functionality for existing persistent solvers
- Cutting stock example to demonstrate functionality

### Legal Acknowledgement

By contributing to this software project, I have read the [contribution guide](https://pyomo.readthedocs.io/en/stable/contribution_guide.html) and agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the BSD license.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

### CODE REVIEW

File: pyomo/solvers/tests/checks/test_gurobi_persistent.py [hunk 1]
```diff
@@ -309,3 +309,23 @@ def _my_callback(cb_m, cb_opt, cb_where):
...
+        m.obj = pe.Objective(expr=-m.x)
+
+        opt = pe.SolverFactory('gurobi_persistent')
+        opt.set_instance(m)
+        opt.solve()
+        self.assertAlmostEqual(m.x.value, 1)
+
+        m.y = pe.Var(within=pe.NonNegativeReals)
+
+        opt.add_column(m, m.y, -2, [m.c], [1])
```
Comment (jsiirola) [inline pyomo/solvers/tests/checks/test_gurobi_persistent.py:327]: I'd like to see this coefficient be something other than `1` (1 is easy to accidentally generate "passes" by having the coefficient actually be omitted).  Could you change this to 2? ... the same comment applies to the example in CPLEX and Xpress

Comment (bknueven) [reply pyomo/solvers/tests/checks/test_gurobi_persistent.py]: Done!

**Comment by bknueven** [comment]
@jsiirola regarding the tests, I had some of these cases in the tests for Xpress but not the others. I've expanded the tests to include all the edge cases you provided across all the persistent solvers.

A few of the edge cases you pointed out would not have been caught, so I added traps for them in `PersistentSolver.add_column`. I also discovered that "`var` is not attached to this model" wasn't caught by `add_var`, so I added a trap for that in `PersistentSolver.add_var` as well.
